File: NeetCode/Two_Pointers/Boats_To_Save_People/boats__to__save__people.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        boatCount = 0
        people.sort()

        left, right = 0, len(people) - 1
        while left <= right:

            if people[left] + people[right] <= limit:

                left += 1
            right -= 1
            boatCount += 1
        return boatCount
        pass


if __name__ == "__main__":
    sol = Solution()
    people = [1, 3, 2, 3, 2]
    limit = 3
    print(sol.numRescueBoats(people, limit))
    people = [5, 1, 4, 2]
    limit = 6
    print(sol.numRescueBoats(people, limit))
    print("Running Solution...")

File: NeetCode/Two_Pointers/Boats_To_Save_People/Boats_To_Save_People.cpp
// Boats_To_Save_People.cpp
//
// Compile locally with:
// clang++ Boats_To_Save_People.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int numRescueBoats(vector<int> &people, int limit)
    {
        sort(people.begin(), people.end());
        int boatCount(0), left(0), right((int)people.size());
        while (left <= right)
        {
            if (people[left] + people[right] <= limit)
            {
                left++;
            }
            right--;
            boatCount++;
        }
        return boatCount;
    }
};

int main()
{
    cout << "Running Boats_To_Save_People.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
